### PR TITLE
[ONNX] Clean up `__init__` in torch.onnx

### LINKED ISSUE #78336: [ONNX] Add `__all__` to `torch.onnx`
### 🚀 The feature, motivation and pitch

The [Public API definition and documentation](https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation) requires torch.onnx define an `__all__` to expose api's explicitly. Examples we should not expose include `torch.onnx.Union` (imported from typing).

- [x] Move class and function definitions out from `__init__.py` into private namespaces, and import them in `__init__.py` to expose the names to prevent internal dependency on anything in `__init__`. (To avoid circular dependency)
- [x] Define `__all__`

### Alternatives

Not define `__all__` and have all objects' `__module__` start with the current submodule. This is harder to do since we at least import from typing.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by justinchuby**
- Move definitions in `__init__` to internal classes and expose them by importing to init (prevent circular dependencies): https://github.com/pytorch/pytorch/wiki/torch.onnx-Namespacing
  - Context classes and enums are moved to `_exporter_states.py`
  - Exceptions are moved to `errors.py`
- Define `__all__` for torch.onnx. https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation
- Moved `utils.__IN_ONNX_EXPORT` to `GLOBALS.in_onnx_export`
- Deprecated `torch.onnx._export`

Precedes #78231

Using this as an aid for finding public functions:

```python
list(filter(lambda x: not x.startswith("_"), torch.onnx.utils.__dict__.keys()))
```

### CODE REVIEW

**Comment by albanD** [comment]
Hey!

I think using the `__all__` and maintaining it is the best way to go as adding/removing a new function does mean updating the doc, etc. So updating the `__all__` shouldn't much more work.
But if that is tricky, you can also populate the `__all__` list programmatically by taking everything from a submodule. It is just a list of strings.

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -1140,31 +1142,32 @@ def _export(
...
-                torch.onnx.ExportTypes.ZIP_ARCHIVE,
-                torch.onnx.ExportTypes.COMPRESSED_ZIP_ARCHIVE,
+                _exporter_states.ExportTypes.ZIP_ARCHIVE,
+                _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE,
             ]:
                 compression = (
                     zipfile.ZIP_DEFLATED
-                    if export_type == torch.onnx.ExportTypes.COMPRESSED_ZIP_ARCHIVE
+                    if export_type
+                    == _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE
```
Comment (BowenBao) [inline torch/onnx/utils.py:1172]: weird format .. is this normal?

Comment (justinchuby) [reply torch/onnx/utils.py:1172]: If could be black thinking line is too long. lintrunner seems to be happy about it.

File: torch/onnx/__init__.py [hunk 2]
```diff
@@ -1,57 +1,84 @@
...
+producer_version = _C_onnx.PRODUCER_VERSION
 
 
 def _export(*args, **kwargs):
-    from torch.onnx import utils
-
-    result = utils._export(*args, **kwargs)
-    return result
+    warnings.warn(
+        "`torch.onnx._export` is deprecated. Please use `export` instead.",
```
Comment (BowenBao) [inline torch/onnx/__init__.py:78]: I'm not sure about this. I agree with the intention. However, we may need to propose the alternative along with message of deprecation.

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -24,13 +24,31 @@
...
+    "unpack_quantized_tensor",
+    "export_to_pretty_string",
+    "unconvertible_ops",
+    "get_ns_op_name_from_custom_op",
+    "register_custom_op_symbolic",
+    "unregister_custom_op_symbolic",
+]
+
 # the flag to tell the user whether it's in the middle of ONNX export or not
 __IN_ONNX_EXPORT = False
```
Comment (BowenBao) [inline torch/onnx/utils.py:53]: Should this be moved to `_exporter_states.py`?

File: torch/onnx/errors.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+"""ONNX exporter exceptions."""
+
+from typing import Optional
+
+from torch.onnx import _constants
+
+__all__ = ["OnnxExporterError", "CheckerError", "UnsupportedOperatorError"]
```
Comment (BowenBao) [inline torch/onnx/errors.py:7]: Is it necessary to define `__all__` for this file?

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -5006,7 +5272,9 @@ def tolist(g, input, dim_val, elem_ty_val):
     # Symbolic functions that need extra context
     # -----------------------------------------------------------------------------
     @staticmethod
-    def device(ctx: torch.onnx.SymbolicContext, g: _C.Graph, *inputs, **kwargs) -> None:
+    def device(
+        ctx: _exporter_states.SymbolicContext, g: _C.Graph, *inputs, **kwargs
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:5276]: To clarify, is `torch.onnx.SymbolicContext` still supported syntax?

I feel the previous version is better. Using class inside private module seems strange for users especially when writing custom symbolic functions on their own. The rest of changes using new `module_name.class_name` format looks good.

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -53,6 +53,272 @@
 #     if it exists, we can write a single function for both overloads.
 #
 
+__all__ = [
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:56]: Why is `__all__` needed for this file but not other `symbolic_opset[version].py`s?

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -2466,7 +2732,7 @@ def _unique2(g, input, sorted, return_inverse, return_counts):
 
 # TODO(justinchuby): Clean up this function generation magic by defining the functions
 # explicitly.
-for k, v in symbolic_helper.cast_pytorch_to_onnx.items():
+for k, v in symbolic_helper.cast_pytorch_to_onnx.items():  # type: ignore[has-type]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:2735]: curious why this change is needed?

File: torch/onnx/__init__.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/__init__.py:78]: The message proposes `export`. Are there cases when we needed to use `torch.onnx._export`? Internally we can always use `utils._export`

File: torch/onnx/utils.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/utils.py:53]: Sure

File: torch/onnx/errors.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/errors.py:7]: Since we have `from typing import Optional`, the CI check will complain w/o all, saying `Optional` doesn't belong to the module

File: torch/onnx/symbolic_opset9.py (hunk 5 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:5276]: Users will still use `torch.onnx.SymbolicContext`. Internally we use `_exporter_states.SymbolicContext` to avoid references to torch.onnx. (for a clean import dependency)

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:56]: In other files we don't import from typing yet. When we do we will need `__all__`

File: torch/onnx/symbolic_opset9.py (hunk 7 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:2735]: I am not sure exactly. mypy complains with this change. My guess is because we added more typing contexts to the code, mypy decided to start reporting errors.

File: torch/onnx/utils.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/utils.py]: I moved it to `GLOBALS.in_onnx_export`

**Comment by justinchuby** [comment]
```
Full list:
2022-06-09T21:12:19.5997082Z # torch.onnx.ExportTypes:
2022-06-09T21:12:19.5997400Z   - Is public: it is inside the module's (`torch.onnx`) `__all__`
2022-06-09T21:12:19.5997958Z   - Does NOT look public: because its `__module__` attribute (`torch.onnx._exporter_states`) is not within the torch library or does not start with the submodule where it is defined (`torch.onnx`)
2022-06-09T21:12:19.5998424Z   - You can do either of these two things to fix this problem:
2022-06-09T21:12:19.5998785Z     - To make it NOT public: remove it from the modules's (`torch.onnx`) `__all__`
2022-06-09T21:12:19.5999230Z     - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.onnx`
2022-06-09T21:12:19.5999538Z # torch.onnx.SymbolicContext:
2022-06-09T21:12:19.6000146Z   - Is public: it is inside the module's (`torch.onnx`) `__all__`
2022-06-09T21:12:19.6000707Z   - Does NOT look public: because its `__module__` attribute (`torch.onnx._exporter_states`) is not within the torch library or does not start with the submodule where it is defined (`torch.onnx`)
2022-06-09T21:12:19.6001181Z   - You can do either of these two things to fix this problem:
2022-06-09T21:12:19.6001556Z     - To make it NOT public: remove it from the modules's (`torch.onnx`) `__all__`
2022-06-09T21:12:19.6001987Z     - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.onnx`
2022-06-09T21:12:19.6002301Z # torch.onnx.symbolic_opset9.lerp:
2022-06-09T21:12:19.6002737Z   - Is NOT public: it is not inside the module's (`torch.onnx.symbolic_opset9`) `__all__`
2022-06-09T21:12:19.6003246Z   - Does look public: it does look public because it follows the rules from the doc above (does not start with `_` and has a proper `__module__`).
2022-06-09T21:12:19.6003656Z   - You can do either of these two things to fix this problem:
2022-06-09T21:12:19.6004049Z     - To make it public: add it from the modules's (`torch.onnx.symbolic_opset9`) `__all__`
2022-06-09T21:12:19.6004420Z     - To make it NOT look public: make its name start with `_`
```

**Comment by justinchuby** [comment]
@albanD `torch.onnx.SymbolicContext` is imported from `torch.onnx._exporter_states` and exported in `torch.onnx`. I wonder if setting its `__module__` is the right thing to do because internal usage will then see a different `__module__` than its original (although it isn't really examined). Do you have any suggestions on this case?

File: torch/onnx/__init__.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/__init__.py]: Users that depend on additional arguments in `_export` that is not in `export`. I'm ok with this. Let's add this to the description.

File: torch/onnx/symbolic_opset9.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:5330]: Would it be clearer to special case this usage with `from _exporter_states import SymbolicContext`, then annotate as `SymbolicContext`, what do you think?

Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:5330]: Sure! We can do that. We just need to make sure `_exporter_states` doesn't take many dependencies to complicate the dependency graph (which is the case now so we are good)

imo we should treat it as exceptional and not use it as an example for future usages

**Review by BowenBao — APPROVED** [review]
LGTM, minor comments inline

File: torch/onnx/__init__.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/__init__.py]: I updated the pr description. Should we consider adding additional arguments to ExportOptions and support them there? This way we have a unified api for exporting.

File: torch/onnx/symbolic_opset9.py (hunk 5 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py]: Done
